full-moon/src/tokenizer.rs

```diff
@@ -955,345 +899,279 @@ impl std::error::Error for TokenizerError {}
-                        token_type: advancement.token_type,
-                    });
-                    if next_is_new_line {
-                        next_is_new_line = false;
-                        position.line += 1;
-                        position.character = 1;
-                    }
+            position.bytes += character.len_utf8();
 
-                    continue;
-                }
+            let end_position = position;
 
-                Ok(None) => {}
+            if next_is_new_line {
+                next_is_new_line = false;
+                position.line += 1;
+                position.character = 1;
+            }
 
-                Err(error) => {
-                    return Err(TokenizerError { error, position });
+            if token_offset == end_position.bytes {
+                tokens.push(start_position, token_type, end_position)?;
+                start_position = position;
+                if let Some((next_token_type, next_token_offset)) = raw_tokens.next() {
+                    token_type = next_token_type;
+                    token_offset = next_token_offset;
+                } else {
+                    break;
                 }
-            };
-        };
-    }
-
-    for _ in 0..1 {
-        // A hack, since `advance!` requires a loop to run.
-        advance!(advance_shebang);
+            }
+        }
     }
 
-    while code.bytes().count() > position.bytes {
-        advance!(advance_whitespace);
-        advance!(advance_comment);
-        advance!(advance_number);
-        advance!(advance_quote);
-        advance!(advance_symbol);
-        advance!(advance_identifier);
-
-        return Err(TokenizerError {
-            error: TokenizerErrorType::UnexpectedToken(
-                code.chars()
-                    .nth(position.character - 1)
-                    .expect("text overflow while giving unexpected token error"),
-            ),
-            position,
-        });
+    if let Some((token_type, token_offset)) = raw_tokens.next() {
+        panic!("(internal full-moon error) Found token {:?} with offset {:?} which is past the end of source", token_type, token_offset);
     }
 
-    tokens.push(Token {
-        start_position: position,
-        end_position: position,
-        token_type: TokenType::Eof,
-    });
-
-    Ok(tokens)
+    Ok(tokens.finish(position))
 }
 
 #[cfg(test)]
 mod tests {
     use crate::tokenizer::*;
     use pretty_assertions::assert_eq;
 
-    macro_rules! test_advancer {
-        ($advancer:ident($code:tt), $result:expr) => {
-            assert_eq!($advancer($code), $result);
+    macro_rules! test_rule {
+        ($rule:ident($code:expr), $result:expr) => {
+            let code: &str = $code;
+            let result: RawToken = $result.into();
+
+            assert_eq!(
+                tokens::$rule(code)
+                    .map_err(|err| from_parser_error(code)(err).error)
+                    .and_then(|v| v),
```

Comment:
Huh?

Reply:
The rules internally return a `RawToken` which is (at least currently) as `Result` with an error of `TokenizerErrorType`, but peg wraps the error into its own result with its own error type. The `map_err` maps the peg error to ours, resulting in a `Result<Result<T, TokenizerErrorType>, TokenizerErrorType>`. The `and_then` flattens this to `Result<T, TokenizerErrorType>` (I'd use [`.flatten`](https://doc.rust-lang.org/std/result/enum.Result.html#method.flatten) if it was stable, which has exactly this implementation).

Reply:
Ah.